enhancements/ingress/logging-api.md

```diff
@@ -0,0 +1,372 @@
+	// HAProxy documentation:
+	// http://cbonte.github.io/haproxy-dconv/2.0/configuration.html#8.2.3
+	//
+	// +optional
+	HttpLogFormat string `json:"httpLogFormat,omitempty"`
+}
+```
+
+The `LoggingDestination` type is a union type:
+
+```go
+// LoggingDestination describes a destination for log messages.
+// +union
+type LoggingDestination struct {
+	// type is the type of destination for logs.  It must be one of the
+	// following:
+	//
+	// * Container
+	//
+	// The ingress operator configures the sidecar container named "logs" on
+	// the ingress controller pod and configures the ingress controller to
+	// write logs to the sidecar.  The logs are then available as container
+	// logs.  The expectation is that the administrator configures a custom
+	// logging solution that reads logs from this sidecar.  Note that using
+	// container logs means that logs may be dropped if the rate of logs
+	// exceeds the container runtime's or the custom logging solution's
+	// capacity.
+	//
+	// * Syslog
+	//
+	// Logs are sent to a syslog endpoint.  The administrator must specify
+	// an endpoint that can receive syslog messages.  The expectation is
+	// that the administrator has configured a custom syslog instance.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type LoggingDestinationType `json:"type"`
+
+	// syslog holds parameters for a syslog endpoint.  Present only if
+	// type is Syslog.
+	//
+	// +optional
+	Syslog *SyslogLoggingDestinationParameters `json:"syslog,omitempty"`
+
+	// container holds parameters for the Container logging destination.
+	// Present only if type is Container.
+	//
+	// +optional
+	Container *ContainerLoggingDestinationParameters `json:"container,omitempty"`
+}
+```
+
+The union discriminator can have either the value `Container` or the value
+`Syslog`:
+
+```go
+// LoggingDestinationType is a type of destination to which to send log
+// messages.
+//
+// +kubebuilder:validation:Enum=Container;Syslog
+type LoggingDestinationType string
+
+const (
+	// Container sends log messages to a sidecar container.
+	ContainerLoggingDestinationType LoggingDestinationType = "Container"
+
+	// Syslog sends log messages to a syslog endpoint.
+	SyslogLoggingDestinationType LoggingDestinationType = "Syslog"
+)
+```
+
+For `Syslog`, an endpoint must be specified, and a facility may be specified (if
+unspecified the default is "local1"); `Container` has no parameters:
+
+```go
+// SyslogLoggingDestinationParameters describes parameters for the Syslog
+// logging destination type.
+type SyslogLoggingDestinationParameters struct {
+	// endpoint identifies the syslog endpoint that receives log messages.
+	// Specify the value using the format using colon-separated host name or
+	// IP address and port number: host:port.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Endpoint string `json:"endpoint"`
```

Comment:
Thinking about future IPv6 requirements, I'm still slightly paranoid about this — is there some standard validation that will account for different address families? Also, is there any need to represent the transport? Is UDP implied? etc.

Reply:
> is there some standard validation that will account for different address families

@ironcladlou what about something like this:
```
// Validate given Endpoint IP address.
func validateEndpoint(ep string) error {
        ip, port, err := net.SplitHostPort(ep)
        // handle error. Optionally validate port number.
        if ip := net.ParseIP(ip); ip != nil {
                return fmt.Errorf("Endpoint must be a valid IP address")
        }
        if ip.IsLoopback() {
                return fmt.Errorf("Endpoint can't be loopback address")
        }
        if ip.IsMulticast() {
                return fmt.Errorf("Endpoint can't be a multicast address")
        }
        if ip.IsLinkLocalUnicast() {
                return fmt.Errorf("Endpoint can't be a link-local unicast address")
        }
        if ip.IsUnspecified() {
                return fmt.Errorf("Endpoint can't be an all zeros address")
        }
        return nil
}
```

Reply:
@ironcladlou,

> Thinking about future IPv6 requirements, I'm still slightly paranoid about this — is there some standard validation that will account for different address families? Also, is there any need to represent the transport? Is UDP implied? etc.

UDP is implied (HAProxy supports UDP and not TCP); I'll document that explicitly.

Does anything else validate addresses in openshift/api as you are describing?  Do we need to do stricter validation than other components do?  I could write a regular expression, but it would be gnarly.

@danehans, that validation looks good, but is adding validation that would require an admission controller a blocker for this enhancement?

Reply:
I don't know re: precedent, I just wanted to make sure we weren't painting ourselves into any corner with regards to endpoint addressing.

Although this now raises another question related to status. What happens if the endpoint is unreachable for any reason (misconfiguration, outage, etc)? Is there some minimum status reporting we should cover here for that case (e.g. logging degraded condition)?

Reply:
I added a section on validation here:  https://github.com/openshift/enhancements/blob/f18621ac562e17bca6377c68e786e757f7a3bdf3/enhancements/ingress/logging-api.md#validation

I propose we validate that the endpoint can be parsed as a host name and IP address or port and possibly probe the endpoint when admitting the ingress controller.  Does that seem reasonable?